// @nx/angular/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsoncEslintParserVersion = exports.oxcProjectRuntimeVersion = exports.jsdomVersion = exports.vitestVersion = exports.jasmineMarblesVersion = exports.typesNodeVersion = exports.jestPresetAngularVersion = exports.lessVersion = exports.tsNodeVersion = exports.autoprefixerVersion = exports.postcssUrlVersion = exports.postcssVersion = exports.tailwindVersion = exports.typescriptEslintVersion = exports.angularEslintVersion = exports.webpackMergeVersion = exports.moduleFederationEnhancedVersion = exports.moduleFederationNodeVersion = exports.browserSyncVersion = exports.typesExpressVersion = exports.expressVersion = exports.typesCorsVersion = exports.corsVersion = exports.tsLibVersion = exports.zoneJsVersion = exports.rxjsVersion = exports.ngrxVersion = exports.ngPackagrVersion = exports.angularDevkitVersion = exports.angularVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
exports.angularVersion = '~21.2.0';
exports.angularDevkitVersion = '~21.2.0';
exports.ngPackagrVersion = '~21.2.0';
exports.ngrxVersion = '^21.0.0';
exports.rxjsVersion = '~7.8.0';
exports.zoneJsVersion = '~0.15.0';
exports.tsLibVersion = '^2.3.0';
exports.corsVersion = '~2.8.5';
exports.typesCorsVersion = '~2.8.5';
exports.expressVersion = '^4.21.2';
exports.typesExpressVersion = '^4.17.21';
exports.browserSyncVersion = '^3.0.0';
exports.moduleFederationNodeVersion = '^2.7.21';
exports.moduleFederationEnhancedVersion = '^2.1.0';
exports.webpackMergeVersion = '^5.8.0';
exports.angularEslintVersion = '^21.2.0';
exports.typescriptEslintVersion = '^7.16.0';
exports.tailwindVersion = '^3.0.2';
exports.postcssVersion = '^8.4.5';
exports.postcssUrlVersion = '~10.1.3';
exports.autoprefixerVersion = '^10.4.0';
exports.tsNodeVersion = '10.9.1';
exports.lessVersion = '^4.3.0';
exports.jestPresetAngularVersion = '~16.0.0';
exports.typesNodeVersion = '20.19.9';
exports.jasmineMarblesVersion = '^0.9.2';
exports.vitestVersion = '^4.0.8';
exports.jsdomVersion = '^27.1.0';
// `@analogjs/vite-plugin-angular` (vitest-analog path) registers an
// `angularVitestPlugin` whose `transform` hook downlevels `@angular/*`
// fesm2022 modules via `vite.transformWithOxc({ target: 'es2016', … })`
// so Zone.js `fakeAsync` can intercept async/await. The lowered code
// emits external `@oxc-project/runtime/helpers/*` imports (oxc default
// `HelperMode = 'Runtime'`). Nothing upstream declares the runtime, so
// we add it explicitly to user projects.
exports.oxcProjectRuntimeVersion = '^0.115.0';
exports.jsoncEslintParserVersion = '^2.1.0';
