// @module-federation/enhanced/dist/src/lib/sharing/tree-shaking/CollectSharedEntryPlugin.js
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });
const require_runtime = require('../../../_virtual/_rolldown/runtime.js');
let path = require("path");
path = require_runtime.__toESM(path);
let fs = require("fs");
fs = require_runtime.__toESM(fs);

//#region src/lib/sharing/tree-shaking/CollectSharedEntryPlugin.ts
const PLUGIN_NAME = "CollectSharedEntryPlugin";
const NODE_MODULES = "node_modules";
const PNPM_STORE = ".pnpm";
function inferPkgVersionFromResource(resource) {
	try {
		const nmIndex = resource.lastIndexOf(NODE_MODULES);
		if (nmIndex === -1) return;
		const after = resource.substring(nmIndex + 12 + 1);
		if (after.startsWith(`${PNPM_STORE}/`)) {
			const encodedPkgWithVersion = after.split(/[\\/]+/)[1];
			if (encodedPkgWithVersion) {
				const encodedPkgWithoutPeers = encodedPkgWithVersion.split("_")[0];
				const versionStart = encodedPkgWithoutPeers.lastIndexOf("@");
				if (versionStart > 0) return encodedPkgWithoutPeers.slice(versionStart + 1);
			}
		}
		const parts = after.split(/[\\/]+/).filter(Boolean);
		if (!parts.length) return;
		let pkgPathParts;
		if (parts[0].startsWith("@") && parts.length >= 2) pkgPathParts = [parts[0], parts[1]];
		else pkgPathParts = [parts[0]];
		const nmBase = resource.substring(0, nmIndex + 12 + 1);
		const pkgJsonPath = path.join(nmBase, ...pkgPathParts, "package.json");
		try {
			const content = fs.readFileSync(pkgJsonPath, "utf-8");
			const v = JSON.parse(content)?.version;
			if (typeof v === "string" && v) return v;
		} catch {}
		return;
	} catch {
		return;
	}
}
var CollectSharedEntryPlugin = class {
	constructor(options) {
		this.name = PLUGIN_NAME;
		this.name = PLUGIN_NAME;
		const { sharedOptions } = options;
		this.sharedOptions = sharedOptions;
		this._collectedEntries = {};
	}
	getData() {
		return this._collectedEntries;
	}
	resetData() {
		this._collectedEntries = {};
	}
	getMatchedSharedOption(request) {
		return this.sharedOptions.find(([sharedName, sharedConfig]) => request === sharedName || request === sharedConfig.import || request === sharedConfig.request || request === sharedConfig.shareKey);
	}
	addRequest(sharedName, resource, version) {
		this._collectedEntries[sharedName] ||= { requests: [] };
		const requests = this._collectedEntries[sharedName].requests;
		if (requests.some(([, existingVersion]) => existingVersion === version)) return;
		requests.push([resource, version]);
	}
	apply(compiler) {
		compiler.hooks.compilation.tap("CollectSharedEntryPlugin", (_compilation, { normalModuleFactory }) => {
			this.resetData();
			normalModuleFactory.hooks.module.tap("CollectSharedEntryPlugin", (module, { resource }) => {
				if (!resource || !("rawRequest" in module)) return module;
				const matchedSharedOption = this.getMatchedSharedOption(module.rawRequest);
				if (!matchedSharedOption) return module;
				const [sharedName, sharedConfig] = matchedSharedOption;
				const configuredVersion = typeof sharedConfig.version === "string" ? sharedConfig.version : void 0;
				const sharedVersion = inferPkgVersionFromResource(resource) || configuredVersion;
				if (!sharedVersion) return module;
				this.addRequest(sharedName, resource, sharedVersion);
				return module;
			});
		});
	}
};

//#endregion
exports.default = CollectSharedEntryPlugin;
