// @rspack/dev-server/dist/servers/SockJSServer.js
"use strict";
/**
 * The following code is modified based on
 * https://github.com/webpack/webpack-dev-server
 *
 * MIT Licensed
 * Author Tobias Koppers @sokra
 * Copyright (c) JS Foundation and other contributors
 * https://github.com/webpack/webpack-dev-server/blob/main/LICENSE
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sockjs = __importStar(require("sockjs"));
const BaseServer_1 = __importDefault(require("./BaseServer"));
// Workaround for sockjs@~0.3.19
// sockjs will remove Origin header, however Origin header is required for checking host.
// See https://github.com/webpack/webpack-dev-server/issues/1604 for more information
{
    const SockjsSession = require('sockjs/lib/transport').Session;
    const { decorateConnection } = SockjsSession.prototype;
    // eslint-disable-next-line func-names
    SockjsSession.prototype.decorateConnection = function (req) {
        decorateConnection.call(this, req);
        const { connection } = this;
        if (connection.headers &&
            !('origin' in connection.headers) &&
            'origin' in req.headers) {
            connection.headers.origin = req.headers.origin;
        }
    };
}
class SockJSServer extends BaseServer_1.default {
    // options has: error (function), debug (function), server (http/s server), path (string)
    constructor(server) {
        super(server);
        const webSocketServerOptions = this.server.options.webSocketServer.options;
        const getSockjsUrl = (options) => {
            if (typeof options.sockjsUrl !== 'undefined') {
                return options.sockjsUrl;
            }
            return '/__webpack_dev_server__/sockjs.bundle.js';
        };
        this.implementation = sockjs.createServer({
            // Use provided up-to-date sockjs-client
            // eslint-disable-next-line camelcase
            sockjs_url: getSockjsUrl(webSocketServerOptions),
            // Default logger is very annoy. Limit useless logs.
            log: (severity, line) => {
                if (severity === 'error') {
                    this.server.logger.error(line);
                }
                else if (severity === 'info') {
                    this.server.logger.log(line);
                }
                else {
                    this.server.logger.debug(line);
                }
            },
        });
        const getPrefix = (options) => {
            if (typeof options.prefix !== 'undefined') {
                return options.prefix;
            }
            return options.path;
        };
        const options = {
            ...webSocketServerOptions,
            prefix: getPrefix(webSocketServerOptions),
        };
        this.implementation.installHandlers(this.server.server, options);
        this.implementation.on('connection', (client) => {
            // Implement the the same API as for `ws`
            client.send = client.write;
            client.terminate = client.close;
            this.clients.push(client);
            client.on('close', () => {
                this.clients.splice(this.clients.indexOf(client), 1);
            });
        });
        this.implementation.close = (callback) => {
            callback();
        };
    }
}
module.exports = SockJSServer;
