// @module-federation/node/dist/src/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_src_plugins_RemotePublicPathRuntimeModule = require('./plugins/RemotePublicPathRuntimeModule.js');
const require_src_plugins_StreamingTargetPlugin = require('./plugins/StreamingTargetPlugin.js');
const require_src_plugins_EntryChunkTrackerPlugin = require('./plugins/EntryChunkTrackerPlugin.js');
const require_src_plugins_NodeFederationPlugin = require('./plugins/NodeFederationPlugin.js');
const require_src_plugins_UniversalFederationPlugin = require('./plugins/UniversalFederationPlugin.js');
const require_src_plugins_ChunkCorrelationPlugin = require('./plugins/ChunkCorrelationPlugin.js');
const require_src_plugins_UniverseEntryChunkTrackerPlugin = require('./plugins/UniverseEntryChunkTrackerPlugin.js');

exports.ChunkCorrelationPlugin = require_src_plugins_ChunkCorrelationPlugin.default;
exports.EntryChunkTrackerPlugin = require_src_plugins_EntryChunkTrackerPlugin.default;
exports.NodeFederationPlugin = require_src_plugins_NodeFederationPlugin.default;
exports.RemotePublicPathPlugin = require_src_plugins_RemotePublicPathRuntimeModule.default;
exports.StreamingTargetPlugin = require_src_plugins_StreamingTargetPlugin.default;
exports.UniversalFederationPlugin = require_src_plugins_UniversalFederationPlugin.default;
exports.UniverseEntryChunkTrackerPlugin = require_src_plugins_UniverseEntryChunkTrackerPlugin.default;
