// @nx/rspack/src/utils/with-nx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withNx = withNx;
const normalize_assets_1 = require("./normalize-assets");
const apply_base_config_1 = require("../plugins/utils/apply-base-config");
const processed = new Set();
/**
 * @param {WithNxOptions} pluginOptions
 * @returns {NxComposableRspackPlugin}
 */
function withNx(pluginOptions = {}) {
    return function makeConfig(config, { options, context }) {
        if (processed.has(config))
            return config;
        (0, apply_base_config_1.applyBaseConfig)({
            ...options,
            ...pluginOptions,
            target: options.target ?? 'web',
            assets: options.assets
                ? options.assets
                : pluginOptions.assets
                    ? (0, normalize_assets_1.normalizeAssets)(pluginOptions.assets, options.root, options.sourceRoot)
                    : [],
            root: context.root,
            projectName: context.projectName,
            targetName: context.targetName,
            configurationName: context.configurationName,
            projectGraph: context.projectGraph,
            useLegacyHtmlPlugin: pluginOptions.useLegacyHtmlPlugin ?? false,
        }, config);
        processed.add(config);
        return config;
    };
}
