// @module-federation/node/dist/src/plugins/NodeFederationPlugin.js
'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });
const require_runtime = require('../../_virtual/_rolldown/runtime.js');
const require_src_plugins_EntryChunkTrackerPlugin = require('./EntryChunkTrackerPlugin.js');
let _module_federation_sdk = require("@module-federation/sdk");

//#region src/plugins/NodeFederationPlugin.ts
const createBundlerLogger = typeof _module_federation_sdk.createInfrastructureLogger === "function" ? _module_federation_sdk.createInfrastructureLogger : _module_federation_sdk.createLogger;
function getRuntimePluginPath() {
	return require.resolve("../runtimePlugin.js");
}
/**
* Class representing a NodeFederationPlugin.
* @class
*/
var NodeFederationPlugin = class {
	/**
	* Create a NodeFederationPlugin.
	* @constructor
	* @param {NodeFederationOptions} options - The options for the NodeFederationPlugin
	* @param {Context} context - The context for the NodeFederationPlugin
	*/
	constructor({ debug, useRuntimePlugin, ...options }, context) {
		this.logger = createBundlerLogger("[ Node Federation Plugin ]");
		this._options = options || {};
		this.context = context || {};
		this.useRuntimePlugin = useRuntimePlugin || false;
	}
	/**
	* Apply method for the NodeFederationPlugin class.
	* @method
	* @param {Compiler} compiler - The webpack compiler.
	*/
	apply(compiler) {
		(0, _module_federation_sdk.bindLoggerToCompiler)(this.logger, compiler, "NodeFederationPlugin");
		const { webpack } = compiler;
		const pluginOptions = this.preparePluginOptions();
		this.updateCompilerOptions(compiler);
		new (this.getModuleFederationPlugin(compiler, webpack))(pluginOptions).apply(compiler);
		new require_src_plugins_EntryChunkTrackerPlugin.default({}).apply(compiler);
	}
	preparePluginOptions() {
		this._options.runtimePlugins = [...this.useRuntimePlugin ? [getRuntimePluginPath()] : [], ...this._options.runtimePlugins || []];
		return {
			...this._options,
			remotes: this._options.remotes || {},
			runtimePlugins: this._options.runtimePlugins,
			dts: this._options.dts ?? false
		};
	}
	updateCompilerOptions(compiler) {
		const chunkFileName = compiler.options?.output?.chunkFilename;
		const uniqueName = compiler?.options?.output?.uniqueName || this._options.name;
		if (typeof chunkFileName === "string" && uniqueName && !chunkFileName.includes(uniqueName)) {
			const suffix = `-[contenthash].js`;
			compiler.options.output.chunkFilename = chunkFileName.replace(".js", suffix);
		}
	}
	getModuleFederationPlugin(compiler, webpack) {
		let ModuleFederationPlugin;
		try {
			return require("@module-federation/enhanced").ModuleFederationPlugin;
		} catch (e) {
			this.logger.error("Can't find @module-federation/enhanced, falling back to webpack ModuleFederationPlugin, this may not work");
			if (this.context.ModuleFederationPlugin) ModuleFederationPlugin = this.context.ModuleFederationPlugin;
			else if (webpack && webpack.container && webpack.container.ModuleFederationPlugin) ModuleFederationPlugin = webpack.container.ModuleFederationPlugin;
			else ModuleFederationPlugin = require("webpack/lib/container/ModuleFederationPlugin");
			return ModuleFederationPlugin;
		}
	}
};

//#endregion
exports.default = NodeFederationPlugin;
