// @module-federation/node/dist/src/plugins/StreamingTargetPlugin.js
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });
const require_src_plugins_CommonJsChunkLoadingPlugin = require('./CommonJsChunkLoadingPlugin.js');

//#region src/plugins/StreamingTargetPlugin.ts
/**
* Class representing a StreamingTargetPlugin
*/
var StreamingTargetPlugin = class {
	/**
	* Create a StreamingTargetPlugin
	* @param {StreamingTargetOptions} options - The options for the plugin
	*/
	constructor(options) {
		this.options = options || {};
	}
	/**
	* Apply the plugin to the compiler
	* @param {Compiler} compiler - The webpack compiler
	*/
	apply(compiler) {
		const { webpack } = compiler;
		compiler.options.output.chunkFormat = "commonjs";
		if (compiler.options.output.enabledLibraryTypes === void 0) compiler.options.output.enabledLibraryTypes = ["commonjs-module"];
		else compiler.options.output.enabledLibraryTypes.push("commonjs-module");
		compiler.options.output.chunkLoading = "async-node";
		compiler.options.output.enabledChunkLoadingTypes = [];
		compiler.options.output.environment = {
			...compiler.options.output.environment,
			dynamicImport: true
		};
		new (webpack?.node?.NodeEnvironmentPlugin || (require("webpack/lib/node/NodeEnvironmentPlugin")))({ infrastructureLogging: compiler.options.infrastructureLogging }).apply(compiler);
		new (webpack?.node?.NodeTargetPlugin || (require("webpack/lib/node/NodeTargetPlugin")))().apply(compiler);
		new require_src_plugins_CommonJsChunkLoadingPlugin.default({
			asyncChunkLoading: true,
			name: this.options.name,
			remotes: this.options.remotes,
			baseURI: compiler.options.output.publicPath,
			promiseBaseURI: this.options.promiseBaseURI,
			debug: this.options.debug
		}).apply(compiler);
	}
};

//#endregion
exports.default = StreamingTargetPlugin;
