// @module-federation/node/dist/src/plugins/StreamingTargetPlugin.mjs
import { __require } from "../../_virtual/_rolldown/runtime.mjs";
import DynamicFilesystemChunkLoadingPlugin from "./CommonJsChunkLoadingPlugin.mjs";

//#region src/plugins/StreamingTargetPlugin.ts
/**
* Class representing a StreamingTargetPlugin
*/
var StreamingTargetPlugin = class {
	/**
	* Create a StreamingTargetPlugin
	* @param {StreamingTargetOptions} options - The options for the plugin
	*/
	constructor(options) {
		this.options = options || {};
	}
	/**
	* Apply the plugin to the compiler
	* @param {Compiler} compiler - The webpack compiler
	*/
	apply(compiler) {
		const { webpack } = compiler;
		compiler.options.output.chunkFormat = "commonjs";
		if (compiler.options.output.enabledLibraryTypes === void 0) compiler.options.output.enabledLibraryTypes = ["commonjs-module"];
		else compiler.options.output.enabledLibraryTypes.push("commonjs-module");
		compiler.options.output.chunkLoading = "async-node";
		compiler.options.output.enabledChunkLoadingTypes = [];
		compiler.options.output.environment = {
			...compiler.options.output.environment,
			dynamicImport: true
		};
		new (webpack?.node?.NodeEnvironmentPlugin || (__require("webpack/lib/node/NodeEnvironmentPlugin")))({ infrastructureLogging: compiler.options.infrastructureLogging }).apply(compiler);
		new (webpack?.node?.NodeTargetPlugin || (__require("webpack/lib/node/NodeTargetPlugin")))().apply(compiler);
		new DynamicFilesystemChunkLoadingPlugin({
			asyncChunkLoading: true,
			name: this.options.name,
			remotes: this.options.remotes,
			baseURI: compiler.options.output.publicPath,
			promiseBaseURI: this.options.promiseBaseURI,
			debug: this.options.debug
		}).apply(compiler);
	}
};

//#endregion
export { StreamingTargetPlugin as default };
